plugins/libplugin-pay.c

```diff
@@ -661,9 +825,86 @@ static struct command_result *payment_finished(struct payment *p)
 			 * return a response twice. */
 			p->cmd = NULL;
 			return command_finished(cmd, ret);
-		} else {
-			return command_fail(p->cmd, JSONRPC2_INVALID_REQUEST,
-					    "Not functional yet");
+		} else if (result.failure == NULL || result.failure->failcode < NODE) {
+			/* This is failing because we have no more routes to try */
+			ret = jsonrpc_stream_fail(cmd, PAY_ROUTE_NOT_FOUND,
+						  NULL);
+			json_add_string(
+			    ret, "message",
+			    tal_fmt(cmd,
+				    "Ran out of routes to try after "
+				    "%d attempt%s: see paystatus",
+				    result.attempts,
+				    result.attempts == 1 ? "" : "s"));
+			return command_finished(cmd, ret);
+
+		}  else {
+			struct payment_result *failure = result.failure;
+			assert(failure!= NULL);
+			ret = jsonrpc_stream_fail(cmd, p->result->code,
+						  failure->message);
+
+			json_add_u64(ret, "id", p->result->id);
+
+			json_add_u32(ret, "failcode", result.failure->failcode);
+			json_add_string(ret, "failcodename",
+					result.failure->failcodename);
+			json_add_num(ret, "code", result.failure->code);
+
+			json_object_start(ret, "data");
+			if (p->bolt11)
+				json_add_string(ret, "bolt11", p->bolt11);
+
+			json_add_hex_talarr(ret, "raw_message",
+					    p->result->raw_message);
+			json_add_num(ret, "created_at", p->start_time.ts.tv_sec);
+			json_add_string(ret, "message", p->result->message);
+			json_add_node_id(ret, "destination", p->destination);
+			json_add_sha256(ret, "payment_hash", p->payment_hash);
+
+			if (result.leafstates & PAYMENT_STEP_SUCCESS) {
+				/* If one sub-payment succeeded then we have
+				 * proof of payment, and the payment is a
+				 * success. */
+				json_add_string(ret, "status", "complete");
```

Comment:
might this be confusing for MPP payments that aren't fully terminated? ie the result comes back 'complete' but we havent technically finished collecting all the parts yet. possible i'm misunderstanding what'd going on here.

Reply:
You're correct in assuming that this can happen with some payments in flight. However this is likely to correspond to the recipient's state much more closely than if we wait for all parts to complete.

If we received at least one preimage it means the recipient has accepted the payment. Assuming they implement MPP correctly it also indicates that all other parts got at least to the recipient. Now if we still have some pending, they are either slow, or even stuck, and that can potentially last for a long time, which is a bad user-experience imho. This is why I prefer showing it as complete as soon as we get the preimage through on at least one part, so recipient and sender agree on the state almost always, rather than the occasional wait for a stuck payment that we know will eventually clear :-)

Reply:
this seems to suggest that we might add a new state to clearly delineate between "your funds have definitely arrived" and "your funds are definitely arriving"

Reply:
Well, Christian is right in that "the recipient has acknowleged payment".  There could be an orthogonal warning ("but we're waiting for some parts to be resolved") which could also apply for the "we've given up" case (assuming that the parts still pending are not sufficient to complete payment, and so the recipient won't take them).

Reply:
I'll add the a field to that effect in a cleanup commit :+1:

Reply:
Oh right this is "send payment" not "receive payment" side of the transaction -- ok makes sense to me. Apologies for the slow response.